# 方案概述
比赛链接：[系统访问风险识别](https://www.datafountain.cn/competitions/580/datasets)

## 赛题介绍
### 赛题背景

随着国家、企业对安全和效率越来越重视，作为安全基础设施之一——统一身份管理(IAM，Identity and Access Management)系统也得到越来越多的关注。 在IAM领域中，其主要安全防护手段是身份鉴别，身份鉴别主要包括账密验证、扫码验证、短信验证、人脸识别及指纹验证等方式。这些身份鉴别方式一般可分为三类，即用户所知(如口令)、所有(如身份证)、特征(如人脸识别及指纹验证)。这些鉴别方式都有其各自的缺点——比如口令，强度高了不容易记住，强度低了又容易丢；又比如人脸识别，做活体验证用户体验不好，静默检测又容易被照片、视频、人脸模型绕过。也因此，在等保2.0中对于三级以上系统要求必须使用两种及以上的鉴别方式对用户进行身份鉴别，以提高身份鉴别的可信度，这种鉴别方式也被称为双因素认证。

对用户来说，双因素认证在一定程度上提高了安全性，但也极大地降低了用户体验。也因此，IAM厂商开始参考用户实体行为分析(UEBA,User and Entity Behavior Analytics)、用户画像等行为分析技术，来探索一种既能确保用户体验，又能提高身份鉴别可信度的方法。而在当前IAM的探索过程中，目前最容易落地的方法是基于规则的行为分析技术，因为它可理解性较高，且容易与其它身份鉴别方式进行联动。
但基于规则的行为分析技术局限性也很明显，首先这种技术是基于经验的，有“宁错杀一千，不放过一个”的特点，其次它也缺少从数据层面来证明是否有人正在尝试窃取或验证非法获取的身份信息，又或者正在使用窃取的身份信息。鉴于此，我们举办这次竞赛，希望各个参赛团队利用竞赛数据和行业知识，建立机器学习、人工智能或数据挖掘模型，来弥补传统方法的缺点，从而解决这一行业难题。

### 赛题任务

本赛题中，参赛团队将基于用户历史的系统访问日志及是否存在风险标记等数据，结合行业知识，构建必要的特征工程，建立机器学习、人工智能或数据挖掘模型，并用该模型预测将来的系统访问是否存在风险。

### 数据简介
本赛题数据是从竹云日志库中抽取某公司一定比例的员工从2022年1月到6月的系统访问日志数据，主要涉及认证日志与风险日志数据。部分字段经过一一对应脱敏处理，供参赛队伍使用。其中认证日志是用户在访问应用系统时产生的行为数据，包括用户名、认证时间、认证城市、接入系统、访问URL等关键信息。

### 数据说明
• 文档说明

| 名称                    | 说明     |
|-----------------------|--------|
| train.csv             | 训练集数据  |
| evaluation_public.csv | 测试集数据  |
| submit_sample.csv     | 提交样例数据 |

• 变量含义说明

| 变量名称                 | 业务含义       | 说明                            |
|----------------------|------------|-------------------------------|
| id                   | 样本ID       |                               |
| user_name            | 用户名        | 若该变量为空，则说明该条日志为用户登录系统前产生      |
| department           | 用户所在部门     |                               |
| ip_transform         | 认证IP(加密后)  | 真实认证IP与加密字符一一对应脱敏处理           |
| device_num_transform | 认证设备号(加密后) | 真实认证设备号与加密字符一一对应脱敏处理          |
| browser_version      | 浏览器版本      |                               |
| browser              | 浏览器        |                               |
| os_type              | 操作系统类型     |                               |
| os_version           | 操作系统版本     |                               |
| op_datetime          | 认证日期时间     |                               |
| ip_type              | IP类型       |                               |
| http_status_code     | HTTP类型码    |                               |
| op_city              | 认证城市       |                               |
| log_system_transform | 接入系统(加密后)  | 真实接入系统与加密字符一一对应脱敏处理           |
| url                  | 访问URL      |                               |
| op_month             | 认证月份       |                               |
| is_risk              | 是否存在风险     | 1：有风险；0：无风险。仅train.csv数据包含该字段 |


## 解决方案
### 机器学习解决方案
#### 特征衍生方案
1. 在特征衍生环节，根据不同类型的可用数据，按照一定的加工逻辑，设计衍生指标。如：可以通过组合“主维度+算子函数+度量+条件维度+时间维度”的逻辑进行指标的逻辑衍生。

#### 模型选用
1. 评分卡模型
2. lightgbm

### 深度学习解决方案
1. 将数据处理为序列数据
2. 使用esim进行建模

# 数据探索性分析（EDA）

In [1]:
import os
import re
import numpy as np
import pandas as pd
import datetime

import matplotlib.pyplot as plt
import seaborn as sns

import scorpyo as sp


pd.set_option('max_rows', 500)

In [2]:
path_project = r'/Users/liliangshan/workspace/python/01_datasets/ccf_system_access_risk_identification'

# path dir
path_row_data = os.path.join(path_project, 'row_data')
path_new_data = os.path.join(path_project, 'new_data')
path_results  = os.path.join(path_project, 'results')

# path row_data
path_train = os.path.join(path_row_data, 'train.csv')
path_test  = os.path.join(path_row_data, 'evaluation_public.csv')
path_sample_submission = os.path.join(path_row_data, 'submit_example.csv')


path_new_train = os.path.join(path_new_data, 'train_lightgbm.csv')
path_new_test  = os.path.join(path_new_data, 'test_lightgbm.csv')

## results
path_output_report = os.path.join(path_results, '01_原始数据探察_20221013.xlsx')

y_label = "is_risk"

In [3]:
df_row_train = sp.read_data(path_train)
df_row_val  = sp.read_data(path_test)

df_row_train['url_sit'] = df_row_train['url'].map(lambda x: x.split('/')[0])
df_row_train['url_page'] = df_row_train['url'].map(lambda x: x.split('/')[1])

df_row_val['url_sit'] = df_row_val['url'].map(lambda x: x.split('/')[0])
df_row_val['url_page'] = df_row_val['url'].map(lambda x: x.split('/')[1])

## 数据描述性统计

In [4]:
_ = sp.excel_label(df_row_train, y=y_label, excel_path=path_output_report, show=True)

------------------------------------------------------------------------------------------
标签分布展示如下:

数据集样本有: 47660个,特征列有:18列, 标签列为: is_risk


label取值  label数量 label占比
0        0    39964  83.85%
1        1     7696  16.15%

id        user_name department ip_transform device_num_transform  \
0   0  guojianping9672         rd    GVhZtW4i1         rqRxAjAL1RYC   
1   1      yangtao1740      sales    l3MuTMPoQ         iKPTa3su50y7   
2   2     wangying9098         rd    4uHWcskWv         1baNbqxMWcCu   

  browser_version  browser os_type os_version          op_datetime ip_type  \
0      firefox_78  firefox     win      win10  2022-01-18 19:10:41      内网   
1       chrome_93   chrome     win      win11  2022-04-01 17:04:00      内网   
2           ie_11       ie     win      win10  2022-03-01 15:53:49      内网   

   http_status_code op_city log_system_transform                   url  \
0               200      成都         2umVQwhiiwNJ          xxx.com/mail   
1               200      深圳         RwHe8Q1R7AlB     business.xxx.com/   
2               200      成都         dwS3cdn15GK4  wpsdoc.xxx.com/kdocs   

  op_month  is_risk           url_sit url_page  
0  2022-01        0           xxx.com     mail  
1  2022-04        0  business.xxx.com           
2  2022-03        0    wpsdoc.xxx.com    kdocs

sheet01.标签分布分析在/Users/liliangshan/workspace/python/01_datasets/ccf_system_access_risk_identification/results/01_原始数据探察_20221013.xlsx中已经存在，我们将对原文件进行覆盖


------------------------------------------------------------------------------------------


In [5]:
sp.excel_detect(df_row_train, sheet_name='02.训练集-数据描述性统计',excel_path=path_output_report)

sheet02.训练集-数据描述性统计在/Users/liliangshan/workspace/python/01_datasets/ccf_system_access_risk_identification/results/01_原始数据探察_20221013.xlsx中已经存在，我们将对原文件进行覆盖


feat_name_row    type   size  missing  unique  zero_ratio  \
0                     id   int64  47660     0.00   47660       0.000   
1              user_name  object  47660     0.08     187       0.000   
2             department  object  47660     0.08       5       0.000   
3           ip_transform  object  47660     0.00    2105       0.000   
4   device_num_transform  object  47660     0.00     844       0.000   
5        browser_version  object  47660     0.00       8       0.000   
6                browser  object  47660     0.00       5       0.000   
7                os_type  object  47660     0.00       2       0.000   
8             os_version  object  47660     0.00       4       0.000   
9            op_datetime  object  47660     0.00   47343       0.000   
10               ip_type  object  47660     0.08       1       0.000   
11      http_status_code   int64  47660     0.00       5       0.000   
12               op_city  object  47660     0.00      14       0.000   
13  log_system_transform  object  47660     0.08       9       0.000   
14                   url  object  47660     0.00      13       0.000   
15              op_month  object  47660     0.00       4       0.000   
16               is_risk   int64  47660     0.00       2       0.839   
17               url_sit  object  47660     0.00       5       0.000   
18              url_page  object  47660     0.00      12       0.000   

    negative_ratio       top1_all_value  top1_all_ratio  \
0              0.0                    0           0.000   
1              NaN        xuxiuying8050           0.007   
2              NaN                   rd           0.654   
3              NaN            w2CfuqTz3           0.007   
4              NaN         O54DfqjlCrhL           0.007   
5              NaN              edge_93           0.372   
6              NaN               chrome           0.431   
7              NaN                  win           0.905   
8              NaN                win10           0.758   
9              NaN  2022-04-02 17:02:24           0.000   
10             NaN                   内网           0.920   
11             0.0                  200           0.948   
12             NaN                   深圳           0.334   
13             NaN         nHrKgKdJ1Mzt           0.322   
14             NaN       xxx.com/github           0.322   
15             NaN              2022-03           0.293   
16             0.0                    0           0.839   
17             NaN              xxx.com           0.603   
18             NaN               github           0.322   

                 mean_or_top1                     std_or_top2  \
0                     23829.5                       13758.401   
1         xuxiuying8050:0.65%             hongchang3029:0.63%   
2                   rd:65.36%                    sales:17.26%   
3             w2CfuqTz3:0.68%                 u9diCFdYZ:0.66%   
4          O54DfqjlCrhL:0.70%              kUa61ygA6gI3:0.68%   
5              edge_93:37.16%                chrome_90:33.38%   
6               chrome:43.15%                     edge:37.16%   
7                  win:90.52%                     macos:9.48%   
8                win10:75.84%                     win7:10.81%   
9   2022-04-02 17:02:24:0.01%       2022-02-16 11:36:12:0.01%   
10                  内网:92.01%                            None   
11                    211.336                          48.817   
12                  深圳:33.38%                       成都:24.54%   
13        nHrKgKdJ1Mzt:32.17%             dwS3cdn15GK4:22.89%   
14      xxx.com/github:32.17%  wpsdoc.xxx.com/download:15.43%   
15             2022-03:29.28%                  2022-04:26.86%   
16                      0.161                           0.368   
17             xxx.com:60.29%           wpsdoc.xxx.com:22.89%   
18              github:32.17%                 download:15.43%   

                  min_or_top3                 1%_or_top4  \
0                         0.

In [6]:
sp.excel_detect(df_row_val, sheet_name='03.测试集-数据描述性统计',excel_path=path_output_report)

sheet03.测试集-数据描述性统计在/Users/liliangshan/workspace/python/01_datasets/ccf_system_access_risk_identification/results/01_原始数据探察_20221013.xlsx中已经存在，我们将对原文件进行覆盖


feat_name_row    type   size  missing  unique  zero_ratio  \
0                     id   int64  25710    0.000   25710         0.0   
1              user_name  object  25710    0.079     187         0.0   
2             department  object  25710    0.079       5         0.0   
3           ip_transform  object  25710    0.000    1192         0.0   
4   device_num_transform  object  25710    0.000     505         0.0   
5        browser_version  object  25710    0.000       8         0.0   
6                browser  object  25710    0.000       5         0.0   
7                os_type  object  25710    0.000       2         0.0   
8             os_version  object  25710    0.000       4         0.0   
9            op_datetime  object  25710    0.000   25542         0.0   
10               ip_type  object  25710    0.079       1         0.0   
11      http_status_code   int64  25710    0.000       5         0.0   
12               op_city  object  25710    0.000      14         0.0   
13  log_system_transform  object  25710    0.079       9         0.0   
14                   url  object  25710    0.000      13         0.0   
15              op_month  object  25710    0.000       2         0.0   
16               url_sit  object  25710    0.000       5         0.0   
17              url_page  object  25710    0.000      12         0.0   

    negative_ratio       top1_all_value  top1_all_ratio  \
0              0.0                    0           0.000   
1              NaN         yuyuzhen3194           0.007   
2              NaN                   rd           0.656   
3              NaN            H0TKapkPL           0.008   
4              NaN         K8Ith9mjHsKo           0.008   
5              NaN              edge_93           0.367   
6              NaN               chrome           0.433   
7              NaN                  win           0.906   
8              NaN                win10           0.765   
9              NaN  2022-06-23 10:51:17           0.000   
10             NaN                   内网           0.921   
11             0.0                  200           0.945   
12             NaN                   深圳           0.334   
13             NaN         nHrKgKdJ1Mzt           0.312   
14             NaN       xxx.com/github           0.312   
15             NaN              2022-06           0.505   
16             NaN              xxx.com           0.587   
17             NaN               github           0.312   

                 mean_or_top1                     std_or_top2  \
0                     12854.5                        7421.982   
1          yuyuzhen3194:0.72%              fengying9449:0.68%   
2                   rd:65.60%                    sales:17.08%   
3             H0TKapkPL:0.76%                 YBCE8ld50:0.70%   
4          K8Ith9mjHsKo:0.79%              3wDqyLqvVCn1:0.72%   
5              edge_93:36.69%                chrome_90:33.55%   
6               chrome:43.31%                     edge:36.69%   
7                  win:90.63%                     macos:9.37%   
8                win10:76.51%                     win7:10.50%   
9   2022-06-23 10:51:17:0.01%       2022-06-01 11:38:12:0.01%   
10                  内网:92.08%                            None   
11                    211.805                          49.329   
12                  深圳:33.42%                       成都:25.36%   
13        nHrKgKdJ1Mzt:31.23%             dwS3cdn15GK4:24.72%   
14      xxx.com/github:31.23%  wpsdoc.xxx.com/download:16.87%   
15             2022-06:50.50%                  2022-05:49.50%   
16             xxx.com:58.69%           wpsdoc.xxx.com:24.72%   
17              github:31.23%                 download:16.87%   

                  min_or_top3                 1%_or_top4  \
0                         0.0                     257.09   
1           pengfan5076:0.67%           linbin8358:0.67%   
2                 other:4.33%           accounting:3.08%   
3             2qWPkWg5V:0.70%            8J3

1. 训练集测试集客户数量都为187人，需要考虑这些人是否是相互包含
2. 测试集中用户登录系统前产生日志数量占比与训练集数量不一致
3. ip和mac地址数量不太一致，存在一个客户多个ip的情况
4. 浏览器、操作系统类型、IP类型、认证城市这种，一般认为同一客户不太会变，考虑从这方面做些特征，以及onehot
5. 说是2022年1月到6月的系统访问日志数据，但op_month只有两个月，很奇怪。

## 数据基础情况分析

客户历史记录数量统计

In [7]:
df_row_train['user_name'].value_counts()

xuxiuying8050       311
hongchang3029       300
tanliu3173          297
liuhong6350         296
lufan2545           296
qiuyan8450          292
liulin3167          291
sunzhiqiang8616     290
yuanwei8501         289
chenghaiyan1579     286
zhanglihua7105      285
zhaoxiang7127       283
pantingting3662     283
pengxia7510         283
jingbo3416          281
maohaiyan4824       280
xiexiaohong5806     280
huanglei6824        278
mayang4022          277
cenglili3725        275
luyan5353           275
lifan7769           273
xujia4357           271
fengying9449        270
gaofeng5184         269
yuandan8814         269
yangyong8917        269
yangtao1740         268
liuyang8834         268
wangshuhua1453      267
heyuhua2679         266
youzhiqiang3249     266
fangxiurong4573     266
huanghui5940        265
liuchunmei3912      265
lichen1456          265
jiangtao2581        262
chenguizhi2238      262
suping3694          260
wanghongmei7436     260
lixia2119           259
wangchang5581   

In [8]:
df_row_val['user_name'].value_counts()

yuyuzhen3194        185
fengying9449        174
pengfan5076         171
linbin8358          171
lijing7913          170
chenxiaohong3284    168
huangning3243       165
liyuzhen4662        162
tanghua6212         162
pantingting3662     161
caili5590           160
fangxiurong4573     159
lufan2545           159
xulanying3873       158
chengjie1656        158
wangxiurong2873     158
caoyu4082           155
gaofeng5184         155
tangguifang4636     154
youzhiqiang3249     153
linbin5576          152
renming5624         152
wanghongmei3888     152
wanghongmei7436     151
wuqian3014          150
shenping7146        150
ligang8428          148
luojun4825          148
yuanwei8501         147
yuanjun5870         147
linyulan9408        147
wangchang5581       146
liguiying8319       146
heyuhua2679         146
genglin9252         146
wangying9098        146
luoxiuzhen8469      146
wangshuhua1453      145
jiangtao2581        144
huting4731          144
duanguiying2657     142
mayang4022      

In [27]:
df_row_train['device_num_transform'].value_counts()

O54DfqjlCrhL    333
kUa61ygA6gI3    322
Rfv57YyO3vny    318
5DmlITfRNR36    316
TzmgdvYq3Kx0    313
               ... 
aUECyyFo55Zy      1
cREgOG9x3d9X      1
NGfeE42d1yHY      1
T4hueKNccs7X      1
A0TLDctT8OUR      1
Name: device_num_transform, Length: 844, dtype: int64

客户交叉情况

In [10]:
print('测试训练集合共同数量：', len(set(df_row_val['user_name'].unique())&set(df_row_val['user_name'].unique())))
print('训练集合-测试集合数量：', len(set(df_row_val['user_name'].unique()) - set(df_row_val['user_name'].unique())))
print('测试集合-训练集合数量：', len(set(df_row_val['user_name'].unique()) - set(df_row_val['user_name'].unique())))

测试训练集合共同数量： 188
训练集合-测试集合数量： 0
测试集合-训练集合数量： 0


客户登录月份

In [11]:
df_row_train['op_month'].unique()

array(['2022-01', '2022-04', '2022-03', '2022-02'], dtype=object)

In [12]:
df_row_val['op_month'].unique()

array(['2022-05', '2022-06'], dtype=object)

一个客户有多个标签

In [13]:
tmp = df_row_train.groupby(['user_name']).agg({y_label:'nunique'}).reset_index()
display(tmp.head())
display(tmp[tmp['is_risk']<2])

user_name  is_risk
0  baojianhua2916        2
1       caili5590        2
2      caohui3132        2
3       caoyu4082        2
4   cendandan2851        2

Empty DataFrame
Columns: [user_name, is_risk]
Index: []

1. 登录人群是同一批
2. 根据mac才可以唯一确定一个登陆流程，

## 单个客户数据分析

In [26]:
df_row_train[['url_page',y_label,'os_version', 'user_name']].groupby(['url_page',y_label]).agg({'count'}).reset_index()

url_page is_risk os_version user_name
                               count     count
0                       0       3396      3396
1                       1        567       567
2      accounting       0        745       745
3      accounting       1        118       118
4      checkingin       0        792       792
5      checkingin       1        156       156
6        download       0       5143      5143
7        download       1       2209      2209
8    getLoginType       0          6         0
9    getLoginType       1        204         0
10  getVerifyCode       0          8         0
11  getVerifyCode       1        201         0
12         github       0      13391     13391
13         github       1       1939      1939
14          kdocs       0       3075      3075
15          kdocs       1        481       481
16      loginAuth       0       3116         0
17      loginAuth       1        275         0
18           mail       0       4036      4036
19           mail       1        607       607
20             oa       0       2725      2725
21             oa       1        413       413
22           task       0       3531      3531
23           task       1        526       526

In [28]:
df_row_train[(df_row_train['device_num_transform']=='TzmgdvYq3Kx0')].sort_values(by='op_datetime')

id   user_name department ip_transform device_num_transform  \
14951  14951  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
20303  20303  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
11444  11444  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
19106  19106  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
18753  18753  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
33542  33542         NaN        NaN    pPgzIf3S4         TzmgdvYq3Kx0   
3257    3257  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
35920  35920  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
35921  35921  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
35922  35922  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
35923  35923  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
35924  35924  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
35925  35925  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
3860    3860  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
46552  46552  tanliu3173         rd    4Ic5OHjIb         TzmgdvYq3Kx0   
14441  14441  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
5357    5357  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
9268    9268  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
20767  20767  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
38767  38767  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
38768  38768  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
38769  38769  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
38770  38770  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
38771  38771  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
38772  38772  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
38773  38773  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
38774  38774  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
15761  15761  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
29961  29961  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
39744  39744  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
39745  39745  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
39746  39746  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
39747  39747  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
39748  39748  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
39749  39749  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
39750  39750  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
39751  39751  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
14268  14268  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
5622    5622  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
29408  29408  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
15947  15947  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
18658  18658  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
17165  17165  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
24300  24300  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
14149  14149  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
11830  11830  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
40396  40396  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
40397  40397  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
40398  40398  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
40399  40399  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
40400  40400  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
40401  40401  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
40402  40402  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx0   
40403  40403  tanliu3173         rd    pPgzIf3S4         TzmgdvYq3Kx

1. 一种模式是客户连续操作download网页，达到一定次数后会显示异常，同时该异常与http_status_code有关
2. kdocs网页，按月进行统计，客户在不同城市时间是否为最晚登录，这个晚指的是凌晨以后

In [38]:
df_row_train[(df_row_train['user_name']=='liuyang8834')
             &(df_row_train['url_page']=='kdocs')].sort_values(by='op_datetime')

id    user_name department ip_transform device_num_transform  \
44592  44592  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
44905  44905  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
15159  15159  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
15151  15151  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
22185  22185  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
45867  45867  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
2560    2560  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
27840  27840  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
47493  47493  liuyang8834         rd    WKIbJdRp7         0L5KlIYwu0Yy   
45193  45193  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
10215  10215  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
46359  46359  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
13818  13818  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
4375    4375  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
26410  26410  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
10700  10700  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
3764    3764  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
9785    9785  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
11129  11129  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
27584  27584  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
16606  16606  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
19645  19645  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
26017  26017  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
21681  21681  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   
12992  12992  liuyang8834         rd    Z3O4VLMqy         0L5KlIYwu0Yy   

      browser_version browser os_type os_version          op_datetime ip_type  \
44592       chrome_77  chrome     win      win10  2022-01-09 12:58:21      内网   
44905       chrome_77  chrome     win      win10  2022-01-13 12:59:38      内网   
15159       chrome_77  chrome     win      win10  2022-01-17 09:09:31      内网   
15151       chrome_77  chrome     win      win10  2022-01-17 19:43:13      内网   
22185       chrome_77  chrome     win      win10  2022-01-20 19:40:52      内网   
45867       chrome_77  chrome     win      win10  2022-01-23 10:21:27      内网   
2560        chrome_77  chrome     win      win10  2022-01-28 14:51:13      内网   
27840       chrome_77  chrome     win      win10  2022-01-30 17:00:39      内网   
47493       chrome_77  chrome     win      win10  2022-01-30 17:31:25      内网   
45193       chrome_77  chrome     win      win10  2022-02-06 07:51:42      内网   
10215       chrome_77  chrome     win      win10  2022-02-08 17:33:20      内网   
46359       chrome_77  chrome     win      win10  2022-02-09 23:12:31      内网   
13818       chrome_77  chrome     win      win10  2022-02-14 15:36:08      内网   
4375        chrome_77  chrome     win      win10  2022-02-14 16:16:16      内网   
26410       chrome_77  chrome     win      win10  2022-02-28 10:23:22      内网   
10700       chrome_77  chrome     win      win10  2022-03-16 16:32:37      内网   
3764        chrome_77  chrome     win      win10  2022-03-17 17:26:03      内网   
9785        chrome_77  chrome     win      win10  2022-03-22 16:36:12      内网   
11129       chrome_77  chrome     win      win10  2022-03-22 17:40:22      内网   
27584       chrome_77  chrome     win      win10  2022-03-23 14:31:52      内网   
16606       chrome_77  chrome     win      win10  2022-04-12 15:17:58      内网   
19645       chrome_77  chrome     win      win10  2022-04-15 17:55:14      内网   
26017       chrome_77  chrome     win      win10  2022-04-21 10:57:42      内网   
21681       chrome_77  chrome     win      win10  2022-04-21 16:23:44      内网   
12992       chrome_77  chrome     win      win10  2022-04-26

In [34]:
df_row_train[(df_row_train['url_page']=='kdocs')].sort_values(by='op_datetime').head(300)

id         user_name  department ip_transform device_num_transform  \
25228  25228     hongchang3029          rd    DhTMwbtS5         5DmlITfRNR36   
22619  22619       yuanwei8501          rd    P5wY5jTM1         LqEZg78nuB9U   
25210  25210   guojianping9672          rd    GVhZtW4i1         rqRxAjAL1RYC   
17439  17439  huangxiufang7404       sales    edI04wZaT         KxJoETnt0nfy   
5380    5380    shixiuzhen2386       sales    WsbzBcKr1         E5Mjdi1Ka59s   
26443  26443        caohui3132          hr    t6UXa0cTT         S05VjSkg47I4   
3547    3547        xielin3822       sales    rY7u3TkaY         X9RHOzThlnFS   
19512  19512      chengjie1656          rd    0MyWTMAmy         PlAdPUvplSW0   
4312    4312          wuyu2799          rd    0g2uVZSsv         4SXmah16O6AH   
10918  10918        yeting6018  accounting    t7KiqE4yx         h8uS3IlwxjSQ   
7659    7659   zhaohongmei2829          rd    4RwkugApp         KqX2PEWxq3Mh   
27808  27808   jiangyuying6195          rd    huLClnEV9         5yApVeSk7drn   
19203  19203       tiankai3780       sales    5DYkWlyEP         2PQLbdJAk4QW   
12585  12585    wangguihua6330          rd    K4F6BcHOA         5SgCho991yJr   
14106  14106         lifan7769          rd    5GqnACy91         6zUqVT1ygOEZ   
2493    2493        xufeng5019          rd    F9a4DlutV         6ypVTOMezZPI   
13864  13864       yanglin6562       sales    eK12oQmm8         GnkVqPSy5nnl   
2129    2129    chenxiulan3024       sales    JQbX5YUgK         aIbV3Qij3kop   
31871  31871        xufeng5019          rd    F9a4DlutV         6ypVTOMezZPI   
2689    2689     liguiying8319          rd    aS11rfWz4         CYUCfieM2Iss   
1718    1718     cendandan2851          rd    l5GL3bnl4         jUG1ecmbP6rI   
9918    9918     maohaiyan4824          rd    3XPunvYti         04F4iKnBDqb6   
21515  21515         lifan7769          rd    5GqnACy91         6zUqVT1ygOEZ   
9756    9756   wangguirong8195       other    7BJUwsqjq         IUM0CBTaOBp6   
22604  22604         liuyu2660       other    G7Shwm0JO         T2H1mGc6fYWL   
23407  23407       yangxin9544          rd    N5GB4gsS7         798WtsqdmzON   
16442  16442  wangzhiqiang5883          rd    dXBMOjrr9         GH5w6zTcZL0R   
20030  20030     hehongxia7416          rd    NGeB9nPD3         swh1mqEhXZFi   
47162  47162     hehongxia7416          rd    PB0qSZdOP         swh1mqEhXZFi   
17738  17738   wanghongmei3888          rd    qPI0fdhj5         t8xfzCnTTmYd   
23315  23315     fanshulan3712       sales    YT75U8tnL         8X5MTrzXSIqA   
28716  28716       taojing6871          rd    7TFdcL1RO         7YM3aaHmgUsf   
15988  15988         liuyu2660       other    G7Shwm0JO         T2H1mGc6fYWL   
2791    2791      wangying7958          rd    TQ1UXfPw0         7rqsgOo0hLOV   
269      269   wangguirong8195       other    7BJUwsqjq         IUM0CBTaOBp6   
30780  30780      huanghui5940          rd    IMcKXlvt0         WdXbb6MizIQV   
28910  28910   tangguifang4636          rd    9kqHtZdjD         b9ADJJLeCoyD   
17866  17866   huangxiulan7221          rd    Q1DZzlONc         CgLOtWGNWc3N   
27175  27175   zhaohongmei2829          rd    4RwkugApp         KqX2PEWxq3Mh   
12111  12111      chenjing7422          rd    db5BFx2dX         rXOriovd7MVW   
3246    3246        mayang4022          rd    pxWp9QpTc         2EmjEhrepKLJ   
22128  22128        liuwen2409          rd    5a7oGUesV         eCi9IsVr1WEJ   
45418  45418  huangxiufang7404       sales    edI04wZaT         KxJoETnt0nfy   
46234  46234        xielin3822       sales    rY7u3TkaY         X9RHOzThlnFS   
43951  43951        jinyun9252  accounting    Im5MDighi         4Jv2QioonsJM   
46304  46304        shiyan6227          rd    9kllcUfJO         1NnRdFzp9NWJ   
44782  44782     wanglihua3720       sales    PPsmrpZx3         oZWobCMy9y7d   
45171  45171   liangshulan8154       sales    yS1GaXpf3         FgFXwvk2F1Wm   
45464  45464        qinlei6092       other    14oeZAnXE         21DcH4wj4tEf   
43946  439

In [15]:
df_row_train[(df_row_train['device_num_transform']=='O54DfqjlCrhL')
             &(df_row_train['url_page']=='github')].sort_values(by='op_datetime')

id      user_name department ip_transform device_num_transform  \
17341  17341  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
31374  31374  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
3102    3102  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
17518  17518  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
27132  27132  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
20405  20405  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
27238  27238  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
6447    6447  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
45145  45145  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
24148  24148  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
25075  25075  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
1625    1625  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
30245  30245  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
47584  47584  xuxiuying8050         rd    J2X3ryA1x         O54DfqjlCrhL   
16700  16700  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
17025  17025  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
7686    7686  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
25327  25327  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
4008    4008  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
1240    1240  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
10757  10757  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
32053  32053  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
5603    5603  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
28510  28510  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
7660    7660  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
6786    6786  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
12639  12639  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
14810  14810  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
46962  46962  xuxiuying8050         rd    f8ysIIKuj         O54DfqjlCrhL   
29258  29258  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
19302  19302  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
29976  29976  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
20862  20862  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
7226    7226  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
44525  44525  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
30292  30292  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
47587  47587  xuxiuying8050         rd    gmmKukwX9         O54DfqjlCrhL   
1288    1288  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
22135  22135  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
21915  21915  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
45817  45817  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
16832  16832  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
30142  30142  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
4062    4062  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
16864  16864  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
9936    9936  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
5563    5563  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
18742  18742  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
16093  16093  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
12499  12499  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
7518    7518  xuxiuying8050         rd    w2CfuqTz3         O54DfqjlCrhL   
25238  25238  xuxiuying8050         rd    w2CfuqTz3       

In [20]:
df_row_train[df_row_train['device_num_transform']=='T4hueKNccs7X'].sort_values(by='op_datetime')

id user_name department ip_transform device_num_transform  \
35475  35475       NaN        NaN    yISDtzxC9         T4hueKNccs7X   

      browser_version browser os_type os_version          op_datetime ip_type  \
35475       chrome_90  chrome     win       win7  2022-04-26 15:59:49     NaN   

       http_status_code op_city log_system_transform                url  \
35475               400      北京                  NaN  xxx.com/loginAuth   

      op_month  is_risk  url_sit   url_page  
35475  2022-04        0  xxx.com  loginAuth

In [15]:
df_row_train[df_row_train['user_name']=='xujie9775'].sort_values(by='op_datetime')

id  user_name department ip_transform device_num_transform  \
3414    3414  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
30415  30415  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
5340    5340  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
24957  24957  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
13560  13560  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
31230  31230  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
4666    4666  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
22959  22959  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
47293  47293  xujie9775      other    9KreK1Eb3         PV9ahGuqwn4t   
3094    3094  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
12552  12552  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
18021  18021  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
13181  13181  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
12800  12800  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
44366  44366  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
25382  25382  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
17902  17902  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
8777    8777  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
24021  24021  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
5581    5581  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
44965  44965  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
46387  46387  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
23014  23014  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
44278  44278  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
26281  26281  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
8231    8231  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
27786  27786  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
23494  23494  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
16485  16485  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
19059  19059  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
26554  26554  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
22929  22929  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
9726    9726  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
24860  24860  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
16127  16127  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
18059  18059  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
740      740  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
24070  24070  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
20113  20113  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
19213  19213  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
32081  32081  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
463      463  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
26307  26307  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
9895    9895  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
9091    9091  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
3191    3191  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
13871  13871  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
12696  12696  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
22822  22822  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
24420  24420  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
46215  46215  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
21180  21180  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
47211  47211  xujie9775      other    jMMbfZik9         PV9ahGuqwn4t   
46169  46169  xujie9775      other    SJwv4mEe7         PV9ahGuqwn4t   
43355  43355  xujie9775      other    SJwv4mEe7   

In [16]:
df_row_train[df_row_train['user_name'].isnull()].sort_values(by='op_datetime')

id user_name department ip_transform device_num_transform  \
35393  35393       NaN        NaN    3Qm3OCoLY         xTOamJ5o9Ugy   
32240  32240       NaN        NaN    C0IxaOdrh         Gk0JoiqhROiD   
33210  33210       NaN        NaN    C4Wb7HV14         lHZQcsid67md   
34373  34373       NaN        NaN    2GkUZeD9D         DIBl5zjCQg9U   
33788  33788       NaN        NaN    5BLwyu5pl         Y0ic4I4cr0UU   
...      ...       ...        ...          ...                  ...   
34535  34535       NaN        NaN    uV1paHVG6         t3ts32NPjtG9   
32905  32905       NaN        NaN    9lcqFeapI         T7mGjWmswm9Z   
35201  35201       NaN        NaN    0mjaEf4SB         8ftsXFm5I1Ej   
33168  33168       NaN        NaN    V1OBTNxYA         6NRAoXZogVDX   
34927  34927       NaN        NaN    3rBYyRzwl         Rz2yXzQm69qw   

      browser_version browser os_type        os_version          op_datetime  \
35393         edge_93    edge     win             win10  2022-01-07 08:43:56   
32240         edge_93    edge     win             win10  2022-01-07 08:47:40   
33210       chrome_90  chrome     win             win10  2022-01-07 09:01:54   
34373       chrome_90  chrome     win             win10  2022-01-07 09:12:24   
33788         edge_93    edge     win             win10  2022-01-07 09:17:51   
...               ...     ...     ...               ...                  ...   
34535         edge_93    edge     win             win10  2022-04-29 18:44:15   
32905       chrome_90  chrome     win             win11  2022-04-29 19:02:23   
35201       safari_13  safari   macos  macos_big_sur_11  2022-04-29 19:17:02   
33168       safari_13  safari   macos  macos_big_sur_11  2022-04-29 19:24:51   
34927       safari_13  safari   macos  macos_big_sur_11  2022-04-29 19:44:42   

      ip_type  http_status_code op_city log_system_transform  \
35393     NaN               400      国外                  NaN   
32240     NaN               200      深圳                  NaN   
33210     NaN               200      深圳                  NaN   
34373     NaN               200      成都                  NaN   
33788     NaN               200      深圳                  NaN   
...       ...               ...     ...                  ...   
34535     NaN               200      深圳                  NaN   
32905     NaN               200      深圳                  NaN   
35201     NaN               200      成都                  NaN   
33168     NaN               200      杭州                  NaN   
34927     NaN               200      深圳                  NaN   

                         url op_month  is_risk  url_sit       url_page  
35393  xxx.com/getVerifyCode  2022-01        1  xxx.com  getVerifyCode  
32240      xxx.com/loginAuth  2022-01        0  xxx.com      loginAuth  
33210      xxx.com/loginAuth  2022-01        0  xxx.com      loginAuth  
34373      xxx.com/loginAuth  2022-01        0  xxx.com      loginAuth  
33788      xxx.com/loginAuth  2022-01        0  xxx.com      loginAuth  
...                      ...      ...      ...      ...            ...  
34535      xxx.com/loginAuth  2022-04        0  xxx.com      loginAuth  
32905      xxx.com/loginAuth  2022-04        0  xxx.com      loginAuth  
35201      xxx.com/loginAuth  2022-04        0  xxx.com      loginAuth  
33168      xxx.com/loginAuth  2022-04        0  xxx.com      loginAuth  
34927      xxx.com/loginAuth  2022-04        0  xxx.com      loginAuth  

[3810 rows x 19 columns]

In [17]:
df_row_train[df_row_train['http_status_code']==404][['is_risk']].value_counts()

is_risk
1          855
0          219
dtype: int64

In [18]:
df_row_train[df_row_train['url_page']=='download'][['is_risk']].value_counts()

is_risk
0          5143
1          2209
dtype: int64

In [19]:
df_row_train[df_row_train['url_page']=='getVerifyCode'][['is_risk']].value_counts()

is_risk
1          201
0            8
dtype: int64

In [20]:
df_row_train['url_page'].unique()

array(['mail', '', 'kdocs', 'github', 'checkingin', 'oa', 'task',
       'accounting', 'loginAuth', 'getVerifyCode', 'getLoginType',
       'download'], dtype=object)

In [21]:
df_row_train[df_row_train['url_page']==''][['is_risk']].value_counts()

is_risk
0          3396
1           567
dtype: int64

1. http_status_code、log_system_transform、url_sit、url_page效果还可以，可以考虑onehot, 或者woe编码
2. 时间类的特征应该是有意义的

# 特征工程

In [22]:
df = pd.concat([df_row_train, df_row_val])
df = df.sort_values(by='op_datetime')

# 认证日期时间
df['op_datetime'] = pd.to_datetime(df['op_datetime'])
# 几点钟
df['hour'] = df['op_datetime'].dt.hour
# 周几
df['dayofweek'] = df['op_datetime'].dt.dayofweek
# 一个月的第几天
df['day'] = df['op_datetime'].dt.day
# 一年的第几月
df['month'] = df['op_datetime'].dt.month

# 用户名-认证时间
df = df.sort_values(by=['user_name', 'op_datetime']).reset_index(drop=True)
# datetime转int时间戳
df['ts'] = df['op_datetime'].values.astype(np.int64) // 10 ** 9
# 按客户进行统计这次认证和上次认证的时间差
df['ts1'] = df.groupby('user_name')['ts'].shift(1)
# 按客户进行统计这次认证和上上次认证的时间差
df['ts2'] = df.groupby('user_name')['ts'].shift(2)
# 按客户进行统计这次认证和上上上次认证的时间差
df['ts3'] = df.groupby('user_name')['ts'].shift(3)
df['ts_diff1'] = df['ts1'] - df['ts']
df['ts_diff2'] = df['ts2'] - df['ts']
df['ts_diff3'] = df['ts3'] - df['ts']

df['hour_sin'] = np.sin(df['hour']/24*2*np.pi)
df['hour_cos'] = np.cos(df['hour']/24*2*np.pi)


In [23]:
un_numeric_f = []
for col in df.columns:
    if not pd.api.types.is_numeric_dtype(df[col]):
        un_numeric_f.append(col)
un_numeric_f

['user_name',
 'department',
 'ip_transform',
 'device_num_transform',
 'browser_version',
 'browser',
 'os_type',
 'os_version',
 'op_datetime',
 'ip_type',
 'op_city',
 'log_system_transform',
 'url',
 'op_month',
 'url_sit',
 'url_page']

In [24]:
df[un_numeric_f].head()

user_name department ip_transform device_num_transform  \
0  baojianhua2916         rd    W4suCwUym         RlZlLWSvh292   
1  baojianhua2916         rd    W4suCwUym         RlZlLWSvh292   
2  baojianhua2916         rd    W4suCwUym         RlZlLWSvh292   
3  baojianhua2916         rd    W4suCwUym         RlZlLWSvh292   
4  baojianhua2916         rd    W4suCwUym         RlZlLWSvh292   

  browser_version browser os_type os_version         op_datetime ip_type  \
0       chrome_90  chrome     win      win10 2022-01-07 18:55:24      内网   
1       chrome_90  chrome     win      win10 2022-01-07 19:43:28      内网   
2       chrome_90  chrome     win      win10 2022-01-10 11:51:39      内网   
3       chrome_90  chrome     win      win10 2022-01-11 10:18:49      内网   
4       chrome_90  chrome     win      win10 2022-01-11 11:56:43      内网   

  op_city log_system_transform             url op_month  url_sit url_page  
0      杭州         nHrKgKdJ1Mzt  xxx.com/github  2022-01  xxx.com   github  
1      杭州         nHrKgKdJ1Mzt  xxx.com/github  2022-01  xxx.com   github  
2      杭州         2umVQwhiiwNJ    xxx.com/mail  2022-01  xxx.com     mail  
3      杭州         nHrKgKdJ1Mzt  xxx.com/github  2022-01  xxx.com   github  
4      杭州         nHrKgKdJ1Mzt  xxx.com/github  2022-01  xxx.com   github

In [25]:
cat_f = ['user_name','department','ip_transform','device_num_transform','browser_version','browser',
 'os_type','os_version','ip_type','op_city','log_system_transform','url','url_sit','url_page']

# 删除时间及类别型变量过多的特征
remove_col = ['op_datetime', 'op_month', 'user_name', 'ip_transform', 'device_num_transform', ]

In [26]:
cat_f = ['user_name', 'department', 'ip_transform', 'device_num_transform', 'browser_version', 'browser',
          'os_type', 'os_version', 'ip_type', 'op_city', 'log_system_transform', 'url',]

for f in cat_f:
    df[f+'_ts_diff_mean'] = df.groupby([f])['ts_diff1'].transform('mean')
    df[f+'_ts_diff_std'] = df.groupby([f])['ts_diff1'].transform('std')
    df[f+'_ts_diff2_mean'] = df.groupby([f])['ts_diff2'].transform('mean')
    df[f+'_ts_diff2_std'] = df.groupby([f])['ts_diff2'].transform('std')
    df[f+'_ts_diff3_mean'] = df.groupby([f])['ts_diff3'].transform('mean')
    df[f+'_ts_diff3_std'] = df.groupby([f])['ts_diff3'].transform('std')


In [27]:
df = df.drop(columns=remove_col)
df.head()

id department browser_version browser os_type os_version ip_type  \
0  29148         rd       chrome_90  chrome     win      win10      内网   
1  21403         rd       chrome_90  chrome     win      win10      内网   
2   2153         rd       chrome_90  chrome     win      win10      内网   
3   6953         rd       chrome_90  chrome     win      win10      内网   
4  12888         rd       chrome_90  chrome     win      win10      内网   

   http_status_code op_city log_system_transform  ...  \
0               200      杭州         nHrKgKdJ1Mzt  ...   
1               200      杭州         nHrKgKdJ1Mzt  ...   
2               200      杭州         2umVQwhiiwNJ  ...   
3               200      杭州         nHrKgKdJ1Mzt  ...   
4               200      杭州         nHrKgKdJ1Mzt  ...   

  log_system_transform_ts_diff2_mean  log_system_transform_ts_diff2_std  \
0                      -95372.721121                      101869.684766   
1                      -95372.721121                      101869.684766   
2                     -101227.892681                      106073.290645   
3                      -95372.721121                      101869.684766   
4                      -95372.721121                      101869.684766   

  log_system_transform_ts_diff3_mean log_system_transform_ts_diff3_std  \
0                     -141655.055306                     122990.238931   
1                     -141655.055306                     122990.238931   
2                     -149558.636775                     126903.136146   
3                     -141655.055306                     122990.238931   
4                     -141655.055306                     122990.238931   

   url_ts_diff_mean  url_ts_diff_std  url_ts_diff2_mean  url_ts_diff2_std  \
0     -48312.637199     72520.154742      -95372.721121     101869.684766   
1     -48312.637199     72520.154742      -95372.721121     101869.684766   
2     -50542.504150     75311.114041     -101227.892681     106073.290645   
3     -48312.637199     72520.154742      -95372.721121     101869.684766   
4     -48312.637199     72520.154742      -95372.721121     101869.684766   

   url_ts_diff3_mean  url_ts_diff3_std  
0     -141655.055306     122990.238931  
1     -141655.055306     122990.238931  
2     -149558.636775     126903.136146  
3     -141655.055306     122990.238931  
4     -141655.055306     122990.238931  

[5 rows x 99 columns]

In [28]:
# 类别型变量one-hot 编码
df = pd.get_dummies(df)

In [29]:
df_train = df[df[y_label].notna()].reset_index(drop=True)
df_test = df[df[y_label].isna()].reset_index(drop=True)

In [30]:
feats = df_train.columns.drop(['id', y_label])
feats

Index(['http_status_code', 'hour', 'dayofweek', 'day', 'month', 'ts', 'ts1',
       'ts2', 'ts3', 'ts_diff1',
       ...
       'url_page_checkingin', 'url_page_download', 'url_page_getLoginType',
       'url_page_getVerifyCode', 'url_page_github', 'url_page_kdocs',
       'url_page_loginAuth', 'url_page_mail', 'url_page_oa', 'url_page_task'],
      dtype='object', length=164)

In [31]:
res = sp.split_data(df_train, y_label)
res.keys()

odict_keys(['train', 'test'])

In [32]:
df_new_train = res['train']
df_new_val   = res['test']

df_new_train.head()

id  http_status_code  is_risk  hour  dayofweek  day  month          ts  \
0   2153               200      0.0    11          0   10      1  1641815499   
1  12888               200      0.0    11          1   11      1  1641902203   
2  13057               200      0.0    11          2   12      1  1641985658   
3   1909               200      0.0    19          2   12      1  1642014025   
4   4307               200      0.0     8          3   13      1  1642064206   

            ts1           ts2  ...  url_page_checkingin  url_page_download  \
0  1.641585e+09  1.641582e+09  ...                    0                  0   
1  1.641896e+09  1.641815e+09  ...                    0                  0   
2  1.641922e+09  1.641902e+09  ...                    0                  0   
3  1.641986e+09  1.641922e+09  ...                    0                  0   
4  1.642014e+09  1.641986e+09  ...                    0                  0   

   url_page_getLoginType  url_page_getVerifyCode  url_page_github  \
0                      0                       0                0   
1                      0                       0                1   
2                      0                       0                1   
3                      0                       0                0   
4                      0                       0                1   

   url_page_kdocs  url_page_loginAuth  url_page_mail  url_page_oa  \
0               0                   0              1            0   
1               0                   0              0            0   
2               0                   0              0            0   
3               0                   0              0            0   
4               0                   0              0            0   

   url_page_task  
0              0  
1              0  
2              0  
3              1  
4              0  

[5 rows x 166 columns]

In [33]:
sp.save_data(df_new_train, path_new_train)
sp.save_data(df_new_val,  path_new_test)

# modeling

In [34]:
params = {
    'learning_rate': 0.05,
    'boosting_type': 'gbdt',
    'objective': 'binary',
    'metric': 'auc',
    'num_leaves': 31,
    'verbose': -1,
    'seed': 2222,
    'n_jobs': -1,
    'feature_fraction': 0.8,
    'bagging_fraction': 0.9,
    'bagging_freq': 4,
    # 'min_child_weight': 10,
}

In [35]:

import time
from sklearn.metrics import roc_auc_score as auc
import lightgbm as lgb
from sklearn.model_selection import StratifiedKFold, KFold

In [36]:

import lightgbm as lgb
import time
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import roc_auc_score as auc
from sklearn.metrics import mean_squared_error as mse
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error as mae
from sklearn.metrics import mean_absolute_percentage_error as mape
from sklearn.model_selection import StratifiedKFold, KFold
from matplotlib.pyplot import plot, show

In [37]:
fold_num = 5
seeds = [2222]
oof = np.zeros(len(df_train))
importance = 0
pred_y = pd.DataFrame()
score = []
for seed in seeds:
    kf = StratifiedKFold(n_splits=fold_num, shuffle=True, random_state=seed)
    # kf = KFold(n_splits=fold_num, shuffle=True, random_state=seed)
    for fold, (train_idx, val_idx) in enumerate(kf.split(df_train[feats], df_train[y_label])):
        print('-----------', fold)
        train = lgb.Dataset(df_train.loc[train_idx, feats],
                            df_train.loc[train_idx, y_label])
        val = lgb.Dataset(df_train.loc[val_idx, feats],
                          df_train.loc[val_idx, y_label])
        model = lgb.train(params, train, valid_sets=[val], 
                          num_boost_round=20000, early_stopping_rounds=100)

        oof[val_idx] += model.predict(df_train.loc[val_idx, feats]) / len(seeds)
        pred_y['fold_%d_seed_%d' % (fold, seed)] = model.predict(df_test[feats])
        importance += model.feature_importance(importance_type='gain') / fold_num
        score.append(auc(df_train.loc[val_idx, y_label], model.predict(df_train.loc[val_idx, feats])))
feats_importance = pd.DataFrame()
feats_importance['name'] = feats
feats_importance['importance'] = importance
display(feats_importance.sort_values('importance', ascending=False)[:30])

df_train['oof'] = oof
display(np.mean(score), np.std(score))

score = np.mean(score)
df_test[y_label] = pred_y.mean(axis=1).values
df_test = df_test.sort_values('id').reset_index(drop=True)

sub = pd.read_csv(path_sample_submission)
sub[y_label] = df_test[y_label].values
sub.to_csv(os.path.join(path_results,time.strftime('lgb_%Y%m%d%H%M_')+'%.5f.csv'%score), index=False)

----------- 0
[1]	valid_0's auc: 0.930003
Training until validation scores don't improve for 100 rounds
[2]	valid_0's auc: 0.930121
[3]	valid_0's auc: 0.930372
[4]	valid_0's auc: 0.930982
[5]	valid_0's auc: 0.932712
[6]	valid_0's auc: 0.932634
[7]	valid_0's auc: 0.93278
[8]	valid_0's auc: 0.93273
[9]	valid_0's auc: 0.931996
[10]	valid_0's auc: 0.932045
[11]	valid_0's auc: 0.932193
[12]	valid_0's auc: 0.932171
[13]	valid_0's auc: 0.93177
[14]	valid_0's auc: 0.932042
[15]	valid_0's auc: 0.932916
[16]	valid_0's auc: 0.933183
[17]	valid_0's auc: 0.933349
[18]	valid_0's auc: 0.933626
[19]	valid_0's auc: 0.933295
[20]	valid_0's auc: 0.933543
[21]	valid_0's auc: 0.933617
[22]	valid_0's auc: 0.933968
[23]	valid_0's auc: 0.934006
[24]	valid_0's auc: 0.934126
[25]	valid_0's auc: 0.934173
[26]	valid_0's auc: 0.934328
[27]	valid_0's auc: 0.93426
[28]	valid_0's auc: 0.934122
[29]	valid_0's auc: 0.934632
[30]	valid_0's auc: 0.934946
[31]	valid_0's auc: 0.934738
[32]	valid_0's auc: 0.934712
[33]	vali

[80]	valid_0's auc: 0.933454
[81]	valid_0's auc: 0.933455
[82]	valid_0's auc: 0.933249
[83]	valid_0's auc: 0.9331
[84]	valid_0's auc: 0.93314
[85]	valid_0's auc: 0.932965
[86]	valid_0's auc: 0.933058
[87]	valid_0's auc: 0.932991
[88]	valid_0's auc: 0.932968
[89]	valid_0's auc: 0.93302
[90]	valid_0's auc: 0.933094
[91]	valid_0's auc: 0.93316
[92]	valid_0's auc: 0.933387
[93]	valid_0's auc: 0.933584
[94]	valid_0's auc: 0.933656
[95]	valid_0's auc: 0.934013
[96]	valid_0's auc: 0.933943
[97]	valid_0's auc: 0.934332
[98]	valid_0's auc: 0.93432
[99]	valid_0's auc: 0.934264
[100]	valid_0's auc: 0.934141
[101]	valid_0's auc: 0.933964
[102]	valid_0's auc: 0.933969
[103]	valid_0's auc: 0.934016
[104]	valid_0's auc: 0.93391
[105]	valid_0's auc: 0.934327
[106]	valid_0's auc: 0.934352
[107]	valid_0's auc: 0.934511
[108]	valid_0's auc: 0.934455
[109]	valid_0's auc: 0.934472
[110]	valid_0's auc: 0.934581
[111]	valid_0's auc: 0.934571
[112]	valid_0's auc: 0.934471
[113]	valid_0's auc: 0.934493
[114]	v

[92]	valid_0's auc: 0.940008
[93]	valid_0's auc: 0.940002
[94]	valid_0's auc: 0.940208
[95]	valid_0's auc: 0.940174
[96]	valid_0's auc: 0.940056
[97]	valid_0's auc: 0.940153
[98]	valid_0's auc: 0.940161
[99]	valid_0's auc: 0.940112
[100]	valid_0's auc: 0.940096
[101]	valid_0's auc: 0.940197
[102]	valid_0's auc: 0.94017
[103]	valid_0's auc: 0.940081
[104]	valid_0's auc: 0.940069
[105]	valid_0's auc: 0.940078
[106]	valid_0's auc: 0.940228
[107]	valid_0's auc: 0.9402
[108]	valid_0's auc: 0.940188
[109]	valid_0's auc: 0.940188
[110]	valid_0's auc: 0.940193
[111]	valid_0's auc: 0.940099
[112]	valid_0's auc: 0.940026
[113]	valid_0's auc: 0.940056
[114]	valid_0's auc: 0.940092
[115]	valid_0's auc: 0.940041
[116]	valid_0's auc: 0.94007
[117]	valid_0's auc: 0.940052
[118]	valid_0's auc: 0.939948
[119]	valid_0's auc: 0.939815
[120]	valid_0's auc: 0.939715
[121]	valid_0's auc: 0.939573
[122]	valid_0's auc: 0.939488
[123]	valid_0's auc: 0.93959
[124]	valid_0's auc: 0.93957
[125]	valid_0's auc: 0.9

[69]	valid_0's auc: 0.934482
[70]	valid_0's auc: 0.934261
[71]	valid_0's auc: 0.934452
[72]	valid_0's auc: 0.934607
[73]	valid_0's auc: 0.934563
[74]	valid_0's auc: 0.934495
[75]	valid_0's auc: 0.934462
[76]	valid_0's auc: 0.934594
[77]	valid_0's auc: 0.934891
[78]	valid_0's auc: 0.93491
[79]	valid_0's auc: 0.934758
[80]	valid_0's auc: 0.934568
[81]	valid_0's auc: 0.934578
[82]	valid_0's auc: 0.934784
[83]	valid_0's auc: 0.934803
[84]	valid_0's auc: 0.934806
[85]	valid_0's auc: 0.934697
[86]	valid_0's auc: 0.934996
[87]	valid_0's auc: 0.935141
[88]	valid_0's auc: 0.935111
[89]	valid_0's auc: 0.935095
[90]	valid_0's auc: 0.934745
[91]	valid_0's auc: 0.934837
[92]	valid_0's auc: 0.93477
[93]	valid_0's auc: 0.934806
[94]	valid_0's auc: 0.934858
[95]	valid_0's auc: 0.935075
[96]	valid_0's auc: 0.935213
[97]	valid_0's auc: 0.93519
[98]	valid_0's auc: 0.935129
[99]	valid_0's auc: 0.934982
[100]	valid_0's auc: 0.935222
[101]	valid_0's auc: 0.935294
[102]	valid_0's auc: 0.935412
[103]	valid_0'

name    importance
13                      hour_cos  47414.185803
0               http_status_code  37308.611818
26     ip_transform_ts_diff_mean  31764.402039
1                           hour  16959.964554
11                      ts_diff3  11722.530132
3                            day   8497.749582
29     ip_transform_ts_diff2_std   7382.382659
2                      dayofweek   7142.379993
27      ip_transform_ts_diff_std   5098.562416
9                       ts_diff1   4488.759857
12                      hour_sin   3866.828915
4                          month   2759.424514
20       department_ts_diff_mean   2514.586785
5                             ts   2291.836922
10                      ts_diff2   2234.869643
6                            ts1   2003.611752
8                            ts3   1365.626581
7                            ts2   1224.613190
80              url_ts_diff_mean    586.769490
31     ip_transform_ts_diff3_std    537.715914
15         user_name_ts_diff_std    512.155758
14        user_name_ts_diff_mean    473.519567
21        department_ts_diff_std    469.520458
19        user_name_ts_diff3_std    389.288452
17        user_name_ts_diff2_std    374.565008
28    ip_transform_ts_diff2_mean    311.999870
30    ip_transform_ts_diff3_mean    272.067216
18       user_name_ts_diff3_mean    157.885244
16       user_name_ts_diff2_mean    150.455236
38  browser_version_ts_diff_mean    148.118272

0.9375430615690267

0.0018823084888451292